File: CS/Serialization/MySerializeHelper.cs
using System.Reflection;
using DevExpress.Utils;
using DevExpress.Utils.Serializing;
using DevExpress.Utils.Serializing.Helpers;
using System;

namespace Serialization
{
	public class MySerializeHelper : SerializeHelper
	{
		public MySerializeHelper(object rootObject)
			: base(rootObject)
		{
		}
		public MySerializeHelper()
			: base()
		{
		}

		protected override IXtraPropertyCollection SerializeObjectCore(object obj, XtraSerializationFlags parentFlags, OptionsLayoutBase options)
		{
			if ( rootObject == null || !(rootObject is MyGridView) )
				return base.SerializeObjectCore(obj, parentFlags, options);

			XtraPropertyInfoCollection properties = base.SerializeObjectCore(obj, parentFlags, options) as XtraPropertyInfoCollection;
			for ( int i = properties.Count - 1; i >= 0; i-- )
			{
				AllowSerializePropertyEventArgs allowArgs = new AllowSerializePropertyEventArgs(obj, properties[i].Name, properties[i].PropertyType, properties[i].Value);
				((MyGridView)rootObject).RaiseAllowSerialzeProperty(allowArgs);
				if ( !allowArgs.SerializeProperty )
					properties.RemoveAt(i);
			}

			SerializeCustomPropertiesEventArgs customArgs = new SerializeCustomPropertiesEventArgs(obj);
			((MyGridView)rootObject).RaiseSerializeCustomProperties(customArgs);
			if ( customArgs.PropertyNames == null || customArgs.PropertyNames.Length == 0 )
				return properties;

			for ( int i = 0; i < customArgs.PropertyNames.Length; i++ )
			{
				PropertyInfo[] propInfos = obj.GetType().GetProperties(BindingFlags.Instance | BindingFlags.Public);
				PropertyInfo property = Array.Find(propInfos, info => info.Name == customArgs.PropertyNames[i]);
				object propertyValue = null;
				if ( property == null )
				{
					CustomPropertyValueEventArgs args = new CustomPropertyValueEventArgs(obj, customArgs.PropertyNames[i]);
					((MyGridView)rootObject).RaiseSetCustomPropertyValue(args);
					if ( args.PropertyValue == null )
						continue;

					propertyValue = args.PropertyValue;
				}
				else
				{
					propertyValue = property.GetValue(obj, null);
					if ( propertyValue == null )
						continue;
				}

				properties.Add(new XtraPropertyInfo(customArgs.PropertyNames[i], propertyValue.GetType(), propertyValue));
			}

			return properties;
		}
	}
}

File: VB/Serialization/MySerializeHelper.vb
Imports Microsoft.VisualBasic
Imports System.Reflection
Imports DevExpress.Utils
Imports DevExpress.Utils.Serializing
Imports DevExpress.Utils.Serializing.Helpers
Imports System

Namespace Serialization
	Public Class MySerializeHelper
		Inherits SerializeHelper
		Public Sub New(ByVal rootObject As Object)
			MyBase.New(rootObject)
		End Sub
		Public Sub New()
			MyBase.New()
		End Sub

		Protected Overrides Function SerializeObjectCore(ByVal obj As Object, ByVal parentFlags As XtraSerializationFlags, ByVal options As OptionsLayoutBase) As IXtraPropertyCollection
			If rootObject Is Nothing OrElse Not(TypeOf rootObject Is MyGridView) Then
				Return MyBase.SerializeObjectCore(obj, parentFlags, options)
			End If

			Dim properties As XtraPropertyInfoCollection = TryCast(MyBase.SerializeObjectCore(obj, parentFlags, options), XtraPropertyInfoCollection)
			For i As Integer = properties.Count - 1 To 0 Step -1
				Dim allowArgs As New AllowSerializePropertyEventArgs(obj, properties(i).Name, properties(i).PropertyType, properties(i).Value)
				CType(rootObject, MyGridView).RaiseAllowSerialzeProperty(allowArgs)
				If (Not allowArgs.SerializeProperty) Then
					properties.RemoveAt(i)
				End If
			Next i

			Dim customArgs As New SerializeCustomPropertiesEventArgs(obj)
			CType(rootObject, MyGridView).RaiseSerializeCustomProperties(customArgs)
			If customArgs.PropertyNames Is Nothing OrElse customArgs.PropertyNames.Length = 0 Then
				Return properties
			End If

			For i As Integer = 0 To customArgs.PropertyNames.Length - 1
				Dim propInfos() As PropertyInfo = obj.GetType().GetProperties(BindingFlags.Instance Or BindingFlags.Public)
				Dim [property] As PropertyInfo = Array.Find(propInfos, Function(info) info.Name = customArgs.PropertyNames(i))
				Dim propertyValue As Object = Nothing
				If [property] Is Nothing Then
					Dim args As New CustomPropertyValueEventArgs(obj, customArgs.PropertyNames(i))
					CType(rootObject, MyGridView).RaiseSetCustomPropertyValue(args)
					If args.PropertyValue Is Nothing Then
						Continue For
					End If

					propertyValue = args.PropertyValue
				Else
					propertyValue = [property].GetValue(obj, Nothing)
					If propertyValue Is Nothing Then
						Continue For
					End If
				End If

				properties.Add(New XtraPropertyInfo(customArgs.PropertyNames(i), propertyValue.GetType(), propertyValue))
			Next i

			Return properties
		End Function
	End Class
End Namespace
